# SQL

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})



In [2]:
def select(query):
    return pd.io.sql.read_sql(sql=text(query), con = engine)

In [3]:
# проверим на дубликаты и пропуски
books= '''SELECT* FROM books'''
display(select(books).head()) 
display(select(books).isna().sum())
display(select(books).duplicated().sum())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

0

In [4]:
# проверим на дубликаты и пропуски
authors= '''SELECT* FROM authors'''
display(select(authors).head())
display(select(authors).isna().sum())
display(select(authors).duplicated().sum())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

author_id    0
author       0
dtype: int64

0

In [5]:
# проверим на дубликаты и пропуски
publishers= '''SELECT* FROM publishers'''
display(select(publishers).head())
display(select(publishers).isna().sum())
display(select(publishers).duplicated().sum())

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

publisher_id    0
publisher       0
dtype: int64

0

In [6]:
# проверим на дубликаты и пропуски
ratings='''SELECT* FROM ratings'''
display(select(ratings).head())
display(select(ratings).isna().sum())
display(select(ratings).duplicated().sum())

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

rating_id    0
book_id      0
username     0
rating       0
dtype: int64

0

In [7]:
# проверим на дубликаты и пропуски
reviews='''SELECT* FROM reviews'''
display(select(reviews).head())
display(select(reviews).isna().sum())
display(select(reviews).duplicated().sum())

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

review_id    0
book_id      0
username     0
text         0
dtype: int64

0

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [8]:
count_book='''
SELECT COUNT(*)
FROM books
WHERE publication_date::date >='01-01-2000'
'''
select(count_book)

count
0    821

**После 1 января 2000 года вышла 821 книга**

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [9]:
rating='''
SELECT a.title,
       COUNT(DISTINCT b.review_id),
       round(AVG(c.rating),2)
FROM books as a
LEFT JOIN reviews as b on a.book_id=b.book_id
LEFT JOIN ratings as c on  a.book_id=c.book_id
GROUP BY a.title'''

select(rating)

title  count  round
0                                         'Salem's Lot      2   3.67
1                   1 000 Places to See Before You Die      1   2.50
2    13 Little Blue Envelopes (Little Blue Envelope...      3   4.67
3    1491: New Revelations of the Americas Before C...      2   4.50
4                                                 1776      4   4.00
..                                                 ...    ...    ...
994           Wyrd Sisters (Discworld  #6; Witches #2)      3   3.67
995                        Xenocide (Ender's Saga  #3)      3   3.40
996                                    Year of Wonders      4   3.20
997                        You Suck (A Love Story  #2)      2   4.50
998  Zen and the Art of Motorcycle Maintenance: An ...      4   3.83

[999 rows x 3 columns]

**Посчитали для каждой книги количество обзоров и среднюю оценку**

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [12]:
top_publisher='''
SELECT a.publisher_id,
       a.publisher,
       count(b.book_id)
FROM publishers as a
inner join books as b on a.publisher_id=b.publisher_id
WHERE num_pages > 50
GROUP BY a.publisher_id
ORDER BY count(b.book_id)DESC
LIMIT 1
'''
select(top_publisher)

publisher_id      publisher  count
0           212  Penguin Books     42

**Больше всего книг выпустило издательство Penguin Books**

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [13]:
author=''' 
WITH best_author as (SELECT a.author,
                     c.book_id,
                     AVG(b.rating) as avg_rating
                     FROM author as a
                     LEFT JOIN books as c ON a.author_id=c.author_id
                     LEFT JOIN ratings as b ON c.book_id=b.book_id
                     GROUP BY a.author, c.book_id
                     HAVING COUNT(b.rating)>=50)
SELECT best_author.author,
       ROUND(AVG(best_author.avg_rating),2) as avg_rating
FROM best_author
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1'''
select(author)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.28

**Автор с самой высокой средней оценкой книг (с 50 и более оценками): J.K. Rowling/Mary GrandPré**

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [14]:
users='''
WITH a as (SELECT count(review_id) as cnt_review
           FROM reviews
           WHERE  username IN (SELECT username
                               FROM ratings
                               GROUP BY username
                               HAVING count(rating_id)>50)
           GROUP BY username)                     
SELECT ROUND(AVG(cnt_review))
FROM a'''
select(users)

round
0   24.0

**Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24**

### Вывод
* После 1 января 2000 года вышла 821 книга.
* Посчитали для каждой книги количество обзоров и среднюю оценку.
* Больше всего книг выпустило издательство Penguin Books.
* Автор с самой высокой средней оценкой книг (с 50 и более оценками): J.K. Rowling/Mary GrandPré.
* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.